Extensions.md

```diff
@@ -16,108 +16,64 @@
-is used when including it into conditional statements. The third argument is the pretty name followed by the description of the extension and the extension directory. The first optional argument is the test directory, which must also contain a CMakeLists.txt file. The seventh argument can be a conditional variable that tells us whether or not to add a third party subdirectory specified by the final extension.
-
-In the lib archive example, we provide all arguments, but your extension may only need the first five and the the test folder. The seventh and eighth arguments must be included in tandem. 
-
-```cmake
-if ( NOT LibArchive_FOUND OR BUILD_LIBARCHIVE )
-	set(BUILD_TP "TRUE")
-endif()
-createExtension(DISABLE_LIBARCHIVE 
-				ARCHIVE-EXTENSIONS 
-				"ARCHIVE EXTENSIONS" 
-				"This Enables libarchive functionality including MergeContent, CompressContent, and (Un)FocusArchiveEntry" 
-				"extensions/libarchive"
-				"${TEST_DIR}/archive-tests"
-				BUILD_TP
-				"thirdparty/libarchive-3.3.2")
+# Extension internals
+Extensions are dynamic libraries loaded at runtime by the agent. An extension makes its 
+capabilities (classes) available to the system through registrars.
+
+``` C++
+// register user-facing classes as
+REGISTER_RESOURCE(InvokeHTTP, "An HTTP client processor which can interact with a configurable HTTP Endpoint. "
+    "The destination URL and HTTP Method are configurable. FlowFile attributes are converted to HTTP headers and the "
+    "FlowFile contents are included as the body of the request (if the HTTP Method is PUT, POST or PATCH).");
+
+// register internal resources as
+REGISTER_INTERNAL_RESOURCE(HTTPClient);
 ```
 
-It is advised that you also add your extension to bootstrap.sh as that is the suggested method of configuring MiNiFi C++
-  
-# C bindings
-To find your classes, you must adhere to a dlsym call back that adheres to the core::ObjectFactory class, like the one below. This object factory will return a list of classes, that we can instantiate through the class loader mechanism. Note that since we are including your code directly into our runtime, we will take care of dlopen and dlsym calls. A map from the class name to the object factory is kept in memory.
+If you decide to put them in a header file, you better make sure that there are no dependencies between extensions,
+as the inclusion of such header from extension "A", will force it to be defined in the including extension "B". This could result
+in shadowing each other's resources. Best to put them in source files.
+
+Some extensions (e.g. `http-curl`) require initialization before use. You need to subclass `Extension` and let the system know by using `REGISTER_EXTENSION`.
 
 ```C++
-class __attribute__((visibility("default"))) HttpCurlObjectFactory : public core::ObjectFactory {
+class HttpCurlExtension : core::extension::Extension {
  public:
```

Comment:
is the http-curl-extension not an extension? do we not initialize/deinitialize the extension? I feel like having two free-floating functions is not really the approach to best model their semantics

the constr/destr => init/deinit correspondance is used in ExtensionInitializer, if we drop that, drop the inheritance and rename HttpCurlExtension to HttpCurlExtensionInitializier and then register a function like
```
static bool httpCurlRegistrar = (ExtensionManager.get().register[what? not Extension, not Init]([] {
  static HttpCurlExtensionInitializer initializer;
}, true);
```
then ExtensionManager cannot for example query the name of the extension it is initializing, because it is not initializing an extension (that would be `extension->init()`) but it just calls a function

Reply:
> is the http-curl-extension not an extension? do we not initialize/deinitialize the extension?

It is, but it's already modeled by the cmake target and the resulting dynamic library. This is about initializing and deinitializing, and this doesn't need a class.

> ExtensionManager cannot [...] query the name of the extension it is initializing

It always has the filename of the library it is loading, that can be used as a name. I don't think we need more than that.

I'm thinking about a model similar to GCC's constructor and destructor function attributes, but done in a portable way, since we control the loading code here. https://stackoverflow.com/a/2053078

Reply:
gcc const/destr (or any load-time running code DllMain) is not viable here as some extensions (e.g. python extension, jni extension) use the configuration object passed onto it to initialize themselves

also there could be interdependencies between extensions, for example loading `http-curl-extension` loads `civet-extension`, so if we only rely on what we load with `dlopen` we miss out on `civet-extension`

Reply:
removed the inheritance

Reply:
regarding the config access in static init, some/most tests link to shared libraries startup time when the config is not yet available